### PR TITLE
fix: remove mocha.opts, adjust tsconfig.json and fix ts issues

### PR DESCRIPTION
**Pull request by rayakoren**
Before submitting a pull request, please ensure the following:

## Description

Please explain the changes you made.

Add a link to the design (if applicable).

### Checklist

- [ ] Code compiles correctly
- [ ] Relevant tests were added (unit / contract / integration)
- [ ] Relevant logs were added
- [ ] Linting run locally successfully
- [ ] All tests pass
- [ ] UA review
- [ ] Design is documented
- [ ] Extended the README / documentation, if necessary
- [ ] Open source is approved

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -146,41 +146,37 @@
...
     "lines": 90,
     "functions": 85,
     "branches": 85,
     "statements": 90,
-    "extension": [
-      ".ts"
-    ],
     "include": [
-      "src/**"
+      "out/src/**"
```
Comment (tal-sapan) [inline package.json:172]: Please check that the coverage works correctly (and fails if you increase it to be too high)

File: src/commands/mtarDeployCommand.ts [hunk 2]
```diff
@@ -54,39 +54,42 @@ export class MtarDeployCommand {
...
         const inputRequest = messages.SELECT_MTA_ARCHIVE;
         const selectionItems: SelectionItem[] = await SelectionItem.getSelectionItems(
           mtarFilesPaths
         );
-        const userSelection: vscode.QuickPickItem = await Utils.displayOptions(
+        const userSelection = await Utils.displayOptions(
           inputRequest,
           selectionItems
         );
+        if (userSelection === undefined) {
```
Comment (tal-sapan) [inline src/commands/mtarDeployCommand.ts:67]: Please check when this use case happens (when the user cancels the QuickPick?)

File: src/logger/settings-changes-handler.ts [hunk 3]
```diff
@@ -24,9 +24,9 @@ export function listenToLogSettingsChanges(context: vscode.ExtensionContext) {
   context.subscriptions.push(
     vscode.workspace.onDidChangeConfiguration(e => {
       if (e.affectsConfiguration(LOGGING_LEVEL_CONFIG_PROP)) {
-        const logLevel: string = vscode.workspace
-          .getConfiguration()
-          .get(LOGGING_LEVEL_CONFIG_PROP);
+        const logLevel: string =
```
Comment (tal-sapan) [inline src/logger/settings-changes-handler.ts:27]: Optional - you could call `getLoggingLevelSetting` instead (so the setting of the default value is only in one place)

File: src/logger/settings.ts [hunk 4]
```diff
@@ -10,14 +10,20 @@ export const SOURCE_TRACKING_CONFIG_PROP = "MTA Tools.sourceLocationTracking";
  * @returns {LogLevel}
  */
 export function getLoggingLevelSetting(): string {
-  return vscode.workspace.getConfiguration().get(LOGGING_LEVEL_CONFIG_PROP);
+  return (
+    vscode.workspace.getConfiguration().get(LOGGING_LEVEL_CONFIG_PROP) ??
```
Comment (tal-sapan) [inline src/logger/settings.ts:14]: Optional - we might want to align the configuration with the `package.json`. For example, change it to `error` here: https://github.com/SAP/vscode-mta-tools/blob/7173a1a6385d2c1148b95e85198eeff682542f5e/package.json#L101

File: tests/commands/mtaBuildCommand.spec.ts [hunk 5]
```diff
@@ -164,7 +165,7 @@ describe("MTA build command unit tests", () => {
         messages.CUSTOM_EVENT_COMMAND_PALETTE
       ])
       .returns();
-    await mtaBuildCommand.mtaBuildCommand(undefined, swa);
+    await mtaBuildCommand.mtaBuildCommand((undefined as unknown) as Uri, swa);
```
Comment (tal-sapan) [inline tests/commands/mtaBuildCommand.spec.ts:168]: You don't need the casting here

File: tests/commands/mtaBuildCommand.spec.ts [hunk 6]
```diff
@@ -195,7 +196,7 @@ describe("MTA build command unit tests", () => {
...
```
Comment (tal-sapan) [inline tests/commands/mtaBuildCommand.spec.ts:199]: You don't need the casting here

File: tests/extension.spec.ts [hunk 7]
```diff
@@ -113,7 +115,10 @@ describe("Extension unit tests", () => {
         .expects("showErrorMessage")
         .withExactArgs(messages.INSTALL_MTA);
       activate(testContext);
-      await addModuleCommand(undefined, undefined);
+      await addModuleCommand(
+        (undefined as unknown) as SWATracker,
+        (undefined as unknown) as Uri
```
Comment (tal-sapan) [inline tests/extension.spec.ts:120]: You don't need the casting here

**Review by tal-sapan — COMMENTED** [review]
See comments

File: src/logger/settings-changes-handler.ts (hunk 3 above)
Comment (rayakoren) [reply src/logger/settings-changes-handler.ts]: Done

File: package.json (hunk 1 above)
Comment (rayakoren) [reply package.json]: Done - it works as expected

File: src/logger/settings.ts (hunk 4 above)
Comment (rayakoren) [reply src/logger/settings.ts]: Done

File: tests/commands/mtaBuildCommand.spec.ts (hunk 5 above)
Comment (rayakoren) [reply tests/commands/mtaBuildCommand.spec.ts]: Done

File: tests/commands/mtaBuildCommand.spec.ts (hunk 6 above)
Comment (rayakoren) [reply tests/commands/mtaBuildCommand.spec.ts]: Done

File: tests/extension.spec.ts (hunk 7 above)
Comment (rayakoren) [reply tests/extension.spec.ts]: Done

**Comment by lidortal** [comment]
Do you plan to switch from tslint to eslint in a separate PR? See #3

File: src/commands/mtaBuildCommand.ts [hunk 8]
```diff
@@ -51,31 +49,35 @@ export class MtaBuildCommand {
...
         const inputRequest = messages.SELECT_PROJECT_DESCRIPTOR;
         const selectionItems: SelectionItem[] = await SelectionItem.getSelectionItems(
           mtaYamlFilesPaths
         );
-        const userSelection: vscode.QuickPickItem = await Utils.displayOptions(
+        const userSelection = await Utils.displayOptions(
           inputRequest,
           selectionItems
         );
+        if (userSelection === undefined) {
```
[the remark this answers is not in the document]
Comment (rayakoren) [reply src/commands/mtaBuildCommand.ts:68]: Yes, when the user cancels QuickPick, strangely it also works without this check ..

**Comment by rayakoren** [comment]
> Do you plan to switch from tslint to eslint in a separate PR? See #3

Yes, this is why it all started... I just do one thing at a time.
The change to eslint will be probably in the next PR

File: src/commands/mtaBuildCommand.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (tal-sapan) [reply src/commands/mtaBuildCommand.ts:68]: From the code it looks like there should be an exception... maybe it's just not displayed anywhere if a command throws an exception.

File: nyc.config.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+  reporter: ["text", "lcov"],
+  include: ["out/src/**"],
+  exclude: ["out/src/logger/**"],
+  // https://reflectoring.io/100-percent-test-coverage/
+  branches: 85,
+  lines: 90,
+  functions: 85,
+  statements: 90,
+  "check-coverage": true,
+  excludeAfterRemap: false,
```
Comment (tal-sapan) [inline nyc.config.js:11]: Please remove this, it's not necessary here. It's done specifically in hana explorer for vscode tests in a vscode environment.

File: nyc.config.js [hunk 10]
```diff
@@ -0,0 +1,13 @@
+module.exports = {
+  reporter: ["text", "lcov"],
+  include: ["out/src/**"],
+  exclude: ["out/src/logger/**"],
+  // https://reflectoring.io/100-percent-test-coverage/
```
Comment (tal-sapan) [inline nyc.config.js:5]: If you don't follow what's described in the link, don't link to it.
Other projects that have this link in the nyc config file actually do have 100% code coverage* as described there.

File: src/commands/mtaBuildCommand.ts [hunk 11]
```diff
@@ -60,6 +66,7 @@ export class MtaBuildCommand {
           selectionItems
         );
         if (userSelection === undefined) {
+          // selection cancled
```
Comment (tal-sapan) [inline src/commands/mtaBuildCommand.ts:69]: canceled

File: src/commands/mtarDeployCommand.ts [hunk 12]
```diff
@@ -65,6 +72,7 @@ export class MtarDeployCommand {
           selectionItems
         );
         if (userSelection === undefined) {
+          // selection cancled
```
Comment (tal-sapan) [inline src/commands/mtarDeployCommand.ts:75]: canceled

File: tests/commands/mtarDeployCommand.spec.ts [hunk 13]
```diff
@@ -230,6 +230,38 @@ describe("Deploy mtar command unit tests", () => {
...
+    selectionItemMock
+      .expects("getSelectionItems")
+      .once()
+      .returns(Promise.resolve());
+    utilsMock
+      .expects("displayOptions")
+      .once()
+      .returns(Promise.resolve(undefined));
+    utilsMock.expects("getConfigFileField").never();
+    utilsMock.expects("getConfigFileField").never();
```
Comment (tal-sapan) [inline tests/commands/mtarDeployCommand.spec.ts:253]: Duplicate line as the one above it

**Review by tal-sapan — COMMENTED** [review]
See comments

File: nyc.config.js (hunk 9 above)
Comment (rayakoren) [reply nyc.config.js:11]: Without it the coverage is not presented

File: nyc.config.js (hunk 10 above)
Comment (rayakoren) [reply nyc.config.js]: Done

File: src/commands/mtaBuildCommand.ts (hunk 11 above)
Comment (rayakoren) [reply src/commands/mtaBuildCommand.ts]: Done

File: src/commands/mtarDeployCommand.ts (hunk 12 above)
Comment (rayakoren) [reply src/commands/mtarDeployCommand.ts]: Done

File: tests/commands/mtarDeployCommand.spec.ts (hunk 13 above)
Comment (rayakoren) [reply tests/commands/mtarDeployCommand.spec.ts]: Done

File: nyc.config.js (hunk 9 above)
Comment (tal-sapan) [reply nyc.config.js:10]: ok

**Review by tal-sapan — APPROVED** [review]

**Pull request merged**